contracts/token/ERC20Channels.sol

```diff
@@ -0,0 +1,303 @@
+    uint256 _challengeTime
+  ) public {
+    require(tokenAddress != address(0));
+    require(_challengeTime >= 0);
+
+    token = ERC20(tokenAddress);
+    challengeTime = _challengeTime;
+  }
+
+  /**
+   * @dev Creates a channel between the msg.sender and the receiver
+   * @param receiver address, the receiver of the channel
+   * @param deposit uint256, the balance taht I want to load in the channel
+   * @param nonce uint8, the nonce number of the channel
+   */
+  function openChannel(
+    address receiver,
+    uint256 deposit,
+    uint8 nonce
+  ) external {
+
+    require(nonce > 0);
+    require(receiver != address(0));
+    require(deposit > 0);
+
+    // Create unique identifier from sender, receiver and current block timestamp
+    bytes32 channelId = getChannelId(msg.sender, receiver, nonce);
+
+    // Check taht teh channel not exist
+    require(channels[channelId].deposit == 0);
+    require(channels[channelId].nonce == 0);
+    require(closingRequests[channelId].closeTime == 0);
+
+    // Store channel information
+    channels[channelId] = Channel({deposit: deposit, nonce: nonce});
+    ChannelCreated(msg.sender, receiver, nonce, deposit);
+
+    // transferFrom deposit from sender to contract
+    // ! needs prior approval from user
+    require(token.transferFrom(msg.sender, address(this), deposit));
+  }
+
+  /**
+   * @dev Starts a close channel request form the sender
+   * @param receiver address, the receiver of the channel
+   * @param nonce uint8, the nonce number of the channel
+   * @param balance uint256, the final balance of teh receiver
+   */
+  function uncooperativeClose(
+    address receiver,
+    uint8 nonce,
+    uint256 balance
+  ) external {
+    bytes32 channelId = getChannelId(msg.sender, receiver, nonce);
+
+    // Check that the closing request dont exist
+    require(closingRequests[channelId].closeTime == 0);
+
+    // Check that the balance is less that the deposited
+    require(balance <= channels[channelId].deposit);
+
+    // Mark channel as closed and create closing request
+    closingRequests[channelId].closeTime = block.timestamp.add(challengeTime);
+    require(closingRequests[channelId].closeTime > block.timestamp);
+    closingRequests[channelId].closingBalance = balance;
+    ChannelCloseRequested(msg.sender, receiver, nonce, balance);
+  }
+
+  /**
+   * @dev Close a channel with the agreement of the sender and receiver
+   * @param receiver address, the receiver of the channel
+   * @param nonce uint8, the nonce number of the channel
+   * @param balanceMsgSig bytes, the signature of the sender
+   * @param closingSig bytes, the signature of the receiver
+   */
+  function cooperativeClose(
+    address receiver,
+    uint8 nonce,
+    uint256 balance,
+    bytes balanceMsgSig,
+    bytes closingSig
+  ) external {
+    // Derive receiver address from signature
+    bytes32 msgHash = keccak256(balanceMsgSig);
+    require(receiver == msgHash.recover(closingSig));
```

Comment:
Also, should we consider adding a variant of this method where the receiver is actually the `msg.sender`, so it does not require an offline signature of `balanceMsgSig`? If I understand correctly, this will typically be called by the `receiver`, so we could simplify the flow for this case.

Reply:
Could be. Or maybe an overload that does not require the extra signature? Either way, since you have more actual experience in using this contract, I'll leave it up to you whether it's worthwhile to implement such method or not.